File: packages/ts/src/zyins/cases/ZeroKnowledgeCaseStorage.ts
/**
 * `ZeroKnowledgeCaseStorage` — the default {@link CaseStorage} adapter.
 *
 * Preserves the ISA platform's E2EE Phase 2 guarantee: the SDK generates a
 * fresh 256-bit AES-GCM key per record, encrypts the payload client-side
 * (with `product` bound as AEAD additional data), posts only the opaque
 * envelope to `/v1/case`, and returns the key as `recallToken` (base64url)
 * — the server never holds key material.
 *
 * Recall reverses the flow: fetch the opaque envelope, decrypt with the
 * caller-supplied `recallToken`. A 404 (absent / expired — by design
 * indistinguishable) surfaces as `null`.
 *
 * Implementation reuses the shipped crypto + wire helpers from `#347`
 * (`encryptCase` / `decryptCase` / `signedCaseRequest`) to avoid forking
 * the cryptographic surface.
 */

import { encryptCase, decryptCase } from '../../account/caseCrypto';
import {
  parseCreatedId,
  parseCaseDetail,
} from '../../account/caseWire';
import {
  signedCaseRequest,
  isSuccess,
  type TCaseRequestContext,
} from '../../account/caseTransport';
import { fromHttpResponse } from '../errors';
import type { CaseRecord, CaseStorage, CaseStoragePutResult } from './CaseStorage';

/** Wire path for the opaque case store. Versioned by the `cases` surface. */
const CASE_PATH = '/v1/case';
/** HTTP status that maps to a `null` record on recall (absent / expired). */
const HTTP_NOT_FOUND = 404;

/**
 * Per-operation context needed by the default zero-knowledge adapter.
 * Mirrors {@link import('../../account/cases').CasesContext} without the
 * viewer-origin field — share-link assembly is the consumer's call,
 * not the adapter's.
 */
export type ZeroKnowledgeCaseStorageContext = TCaseRequestContext;

/**
 * Default zero-knowledge implementation of {@link CaseStorage}. Constructed
 * with the same signed-request context the legacy `account.cases` surface
 * uses; the parent `Isa` wires this automatically when no override is
 * supplied on {@link import('../isaOptions').IsaCreateOptions.caseStorage}.
 */
export class ZeroKnowledgeCaseStorage implements CaseStorage {
  constructor(private readonly contextOnce: () => ZeroKnowledgeCaseStorageContext) {}

  async put(record: CaseRecord): Promise<CaseStoragePutResult> {
    if (!record || typeof record.product !== 'string' || record.product.length === 0) {
      throw new Error('ZeroKnowledgeCaseStorage: put requires a non-empty product');
    }
    if (record.payload === undefined) {
      throw new Error('ZeroKnowledgeCaseStorage: put requires a payload');
    }
    const ctx = this.contextOnce();
    const { envelope, keyFragment } = await encryptCase(record.product, record.payload);
    const body = JSON.stringify({ product: record.product, ...envelope });
    const response = await signedCaseRequest(
      { method: 'POST', path: CASE_PATH, body, idempotencyOp: 'cases_create' },
      ctx,
    );
    if (!isSuccess(response.status)) {
      throw fromHttpResponse(response.status, response.body);
    }
    const id = parseCreatedId(response.body);
    return { id, recallToken: keyFragment };
  }

  async get(id: string, recallToken?: string): Promise<CaseRecord | null> {
    if (typeof id !== 'string' || id.length === 0) {
      throw new Error('ZeroKnowledgeCaseStorage: get requires a non-empty id');
    }
    if (typeof recallToken !== 'string' || recallToken.length === 0) {
      throw new Error(
        'ZeroKnowledgeCaseStorage: get requires the recallToken returned from put — ' +
          'records are opaque ciphertext without the per-record key',
      );
    }
    const ctx = this.contextOnce();
    const path = `${CASE_PATH}/${encodeURIComponent(id)}`;
    const response = await signedCaseRequest({ method: 'GET', path, body: '' }, ctx);
    if (response.status === HTTP_NOT_FOUND) return null;
    if (!isSuccess(response.status)) {
      throw fromHttpResponse(response.status, response.body);
    }
    const { product, envelope } = parseCaseDetail(response.body);
    const payload = await decryptCase(product, envelope, recallToken);
    return { product, payload };
  }
}

File: packages/csharp/src/Zyins/Cases/ZeroKnowledgeCaseStorage.cs
// Default zero-knowledge ICaseStorage implementation. Locked by
// PR #361 — payloads are sealed client-side under a per-record
// AES-GCM-256 key; only ciphertext leaves the SDK. The recall token
// is the base64url-encoded encryption key, returned to the caller
// and required for retrieval. The server stores the ciphertext at
// /v1/case but never the key.
//
// Cryptographic envelope:
//   - key  : 32 random bytes (AES-256), from RandomNumberGenerator
//   - nonce: 12 random bytes per record (AES-GCM standard)
//   - tag  : 16 bytes (AES-GCM standard)
//   - wire : { id, nonce_b64, ciphertext_b64, tag_b64 }
//
// Mirrors packages/ts/src/zyins/zeroKnowledgeCaseStorage.ts (PR #347).
//
// Platform note: AesGcm is in-box on net6.0+ but not on
// netstandard2.0 (only via Cng on Windows). On netstandard2.0 the
// Default singleton throws PlatformNotSupportedException at first use
// rather than silently degrading. Consumers on the legacy target who
// need zero-knowledge storage must implement ICaseStorage themselves.
using System;
using System.Security.Cryptography;
using System.Threading;
using System.Threading.Tasks;

namespace Isa.Sdk.Zyins.Cases;

/// <summary>
/// Default <see cref="ICaseStorage"/> — encrypts case payloads
/// client-side under a per-record key (AES-GCM-256); the server holds
/// only ciphertext. The recall token is the base64url-encoded key,
/// returned to the caller in <see cref="PutResult.RecallToken"/> and
/// required by <see cref="GetAsync"/> to decrypt the response.
///
/// The default singleton is unwired — calling its methods throws
/// <see cref="NotSupportedException"/> until a transport is attached
/// via <see cref="WithTransport"/>. The IsaOptions resolver substitutes
/// the wired instance built from the active <c>ZyInsClient</c>.
/// </summary>
public sealed class ZeroKnowledgeCaseStorage : ICaseStorage
{
    private const int KeyBytes = 32;     // AES-256
    private const int NonceBytes = 12;   // AES-GCM standard
    private const int TagBytes = 16;     // AES-GCM standard

    private readonly ICaseStorageTransport? _transport;

    private ZeroKnowledgeCaseStorage(ICaseStorageTransport? transport)
    {
        _transport = transport;
    }

    /// <summary>
    /// Process-wide default. Unwired — calling its methods throws
    /// <see cref="NotSupportedException"/>. The options resolver
    /// substitutes a wired instance per <see cref="ZyInsClient"/>.
    /// </summary>
    public static ZeroKnowledgeCaseStorage Default { get; } = new ZeroKnowledgeCaseStorage(transport: null);

    /// <summary>
    /// Construct a wired instance backed by <paramref name="transport"/>.
    /// </summary>
    public static ZeroKnowledgeCaseStorage WithTransport(ICaseStorageTransport transport)
    {
        if (transport is null) throw new ArgumentNullException(nameof(transport));
        return new ZeroKnowledgeCaseStorage(transport);
    }

    /// <inheritdoc />
#if NET8_0_OR_GREATER
    public async Task<PutResult> PutAsync(CaseRecord record, CancellationToken ct = default)
    {
        if (record is null) throw new ArgumentNullException(nameof(record));
        var transport = RequireTransport();
        var key = new byte[KeyBytes];
        RandomNumberGenerator.Fill(key);
        try
        {
            var nonce = new byte[NonceBytes];
            RandomNumberGenerator.Fill(nonce);
            var ciphertext = new byte[record.Payload.Length];
            var tag = new byte[TagBytes];

            using (var gcm = new AesGcm(key, TagBytes))
            {
                gcm.Encrypt(nonce, record.Payload, ciphertext, tag);
            }

            var envelope = new CipherEnvelope(
                Id: record.Id,
                NonceB64: Convert.ToBase64String(nonce),
                CiphertextB64: Convert.ToBase64String(ciphertext),
                TagB64: Convert.ToBase64String(tag));

            // Encode the recall token before zeroing key; the server only
            // ever sees ciphertext, never the key.
            var recall = Base64Url.Encode(key);
            await transport.PutCiphertextAsync(envelope, ct).ConfigureAwait(false);
            // The id the caller addresses the record by is the one they
            // supplied; the server's echoed id is not trusted to redefine
            // the caller's local identity.
            return new PutResult(Id: record.Id, RecallToken: recall);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(key);
        }
    }
#else
    public Task<PutResult> PutAsync(CaseRecord record, CancellationToken ct = default)
    {
        if (record is null) throw new ArgumentNullException(nameof(record));
        _ = RequireTransport();
        return Task.FromException<PutResult>(NotSupportedOnLegacyTarget());
    }
#endif

    /// <inheritdoc />
#if NET8_0_OR_GREATER
    public async Task<CaseRecord?> GetAsync(string id, string? recallToken = null, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("ZeroKnowledgeCaseStorage.GetAsync: id must be non-empty", nameof(id));
        }
        if (string.IsNullOrWhiteSpace(recallToken))
        {
            throw new ArgumentException(
                "ZeroKnowledgeCaseStorage.GetAsync: recallToken is required (the key never leaves the SDK)",
                nameof(recallToken));
        }
        var transport = RequireTransport();
        var envelope = await transport.GetCiphertextAsync(id, ct).ConfigureAwait(false);
        if (envelope is null) return null;

        var key = Base64Url.Decode(recallToken!);
        try
        {
            if (key.Length != KeyBytes)
            {
                throw new CryptographicException(
                    $"ZeroKnowledgeCaseStorage.GetAsync: recallToken decodes to {key.Length} bytes; expected {KeyBytes}");
            }
            var nonce = Convert.FromBase64String(envelope.NonceB64);
            var ciphertext = Convert.FromBase64String(envelope.CiphertextB64);
            var tag = Convert.FromBase64String(envelope.TagB64);
            // AES-GCM authenticates the tag during Decrypt; a truncated or
            // forged tag raises CryptographicException. Reject any tag that
            // is not exactly the standard 16 bytes up front so a malicious
            // server cannot probe shortened-tag behavior.
            if (tag.Length != TagBytes)
            {
                throw new CryptographicException(
                    $"ZeroKnowledgeCaseStorage.GetAsync: authentication tag is {tag.Length} bytes; expected {TagBytes}");
            }
            var plaintext = new byte[ciphertext.Length];

            using (var gcm = new AesGcm(key, TagBytes))
            {
                gcm.Decrypt(nonce, ciphertext, tag, plaintext);
            }

            // Address the record by the id the caller requested, not the
            // server-echoed envelope.Id: a tampering server must not be able
            // to rebind decrypted plaintext to a different identity.
            return new CaseRecord(id, plaintext);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(key);
        }
    }
#else
    public Task<CaseRecord?> GetAsync(string id, string? recallToken = null, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("ZeroKnowledgeCaseStorage.GetAsync: id must be non-empty", nameof(id));
        }
        if (string.IsNullOrWhiteSpace(recallToken))
        {
            throw new ArgumentException(
                "ZeroKnowledgeCaseStorage.GetAsync: recallToken is required (the key never leaves the SDK)",
                nameof(recallToken));
        }
        _ = RequireTransport();
        return Task.FromException<CaseRecord?>(NotSupportedOnLegacyTarget());
    }
#endif

    /// <inheritdoc />
    public Task DeleteAsync(string id, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("ZeroKnowledgeCaseStorage.DeleteAsync: id must be non-empty", nameof(id));
        }
        var transport = RequireTransport();
        return transport.DeleteAsync(id, ct);
    }

    private ICaseStorageTransport RequireTransport()
    {
        if (_transport is null)
        {
            throw new NotSupportedException(
                "ZeroKnowledgeCaseStorage.Default is unwired. Construct with WithTransport(...) " +
                "or pass a custom ICaseStorage via IsaOptions.CaseStorage.");
        }
        return _transport;
    }

#if !NET8_0_OR_GREATER
    private static PlatformNotSupportedException NotSupportedOnLegacyTarget() =>
        new PlatformNotSupportedException(
            "ZeroKnowledgeCaseStorage: AES-GCM is not available on netstandard2.0. " +
            "Target net6.0+ or supply a custom ICaseStorage via IsaOptions.CaseStorage.");
#endif
}

/// <summary>
/// Cipher-envelope wire shape. The server stores this verbatim; the
/// key never leaves the SDK.
/// </summary>
public sealed record CipherEnvelope(string Id, string NonceB64, string CiphertextB64, string TagB64);

/// <summary>
/// Transport seam used by <see cref="ZeroKnowledgeCaseStorage"/> to
/// reach the <c>/v1/case</c> surface. Injecting the transport keeps
/// the storage class free of HTTP plumbing and trivially testable
/// (in-memory fakes implement this without HTTP).
/// </summary>
public interface ICaseStorageTransport
{
    /// <summary>POST the ciphertext envelope to <c>/v1/case</c>.</summary>
    Task<CipherEnvelope> PutCiphertextAsync(CipherEnvelope envelope, CancellationToken ct);
    /// <summary>GET the ciphertext envelope from <c>/v1/case/{id}</c>; null when not found.</summary>
    Task<CipherEnvelope?> GetCiphertextAsync(string id, CancellationToken ct);
    /// <summary>DELETE <c>/v1/case/{id}</c>.</summary>
    Task DeleteAsync(string id, CancellationToken ct);
}

internal static class Base64Url
{
    internal static string Encode(byte[] input)
    {
        if (input is null) throw new ArgumentNullException(nameof(input));
        var b64 = Convert.ToBase64String(input);
        return b64.TrimEnd('=').Replace('+', '-').Replace('/', '_');
    }

    internal static byte[] Decode(string input)
    {
        if (input is null) throw new ArgumentNullException(nameof(input));
        var b64 = input.Replace('-', '+').Replace('_', '/');
        switch (b64.Length % 4)
        {
            case 2: b64 += "=="; break;
            case 3: b64 += "="; break;
            case 0: break;
            default: throw new FormatException("Base64Url.Decode: invalid input length");
        }
        return Convert.FromBase64String(b64);
    }
}
